lib/seahorn/Houdini.cc

```diff
@@ -0,0 +1,292 @@
+
+  /*
+   * Extract all predicates in an expression
+   */
+  template<typename OutputIterator>
+  void Houdini::get_all_pred_apps (Expr e, HornClauseDB &db, OutputIterator out)
+  {filter (e, IsPredApp(db), out);}
+
+  bool Houdini::runOnModule (Module &M)
+  {
+    HornifyModule &hm = getAnalysis<HornifyModule> ();
+
+    //load the Horn clause database
+    auto &db = hm.getHornClauseDB ();
+    db.buildIndexes ();
+
+    runHoudini(db, hm);
+
+    return false;
+  }
+
+  void Houdini::getAnalysisUsage (AnalysisUsage &AU) const
+  {
+    AU.addRequired<HornifyModule> ();
+    AU.setPreservesAll();
+  }
+
+  /*
+   * Use template to convert a predicate to candidate
+   */
+  Expr Houdini::relToCand(Expr fdecl)
+  {
+	ExprVector bvars;
+	ExprVector bins;   // a vector of LT expressions
+	Expr cand = NULL;
+
+	int bvar_count = 0;
+	unsigned i = 0;
+	for (i=0; i < bind::domainSz(fdecl); i++)
+	{
+		// if its type is INT
+		if (isOpX<INT_TY>(bind::domainTy(fdecl, i)))
+		{
+			// what is efac?
+			Expr bvar = bind::bvar (i, mk<INT_TY>(bind::domainTy(fdecl, i)->efac())); //the id of bvar is the same as related arg index
+			bvars.push_back(bvar);
+			bvar_count ++;
+		}
+	}
+
+	//What if there's no bvar?
+	if(bvar_count == 0)
+	{
+		cand = mk<TRUE>(fdecl->efac());
+	}
+	// if there is only one bvar, create a int constant and make an lt expr
+	else if(bvar_count == 1)
+	{
+		Expr zero = mkTerm<mpz_class> (0, fdecl->efac());
+		cand = mk<LT>(bvars[0], zero);
+	}
+	// if there are more than two bvars, make an lt expr
+	else if(bvar_count == 2)
+	{
+		Expr lt1 = mk<LT>(bvars[0], bvars[1]);
+		Expr lt2 = mk<LT>(bvars[1], bvars[0]);
+		bins.push_back(lt1);
+		bins.push_back(lt2);
+
+		cand = mknary<AND>(bins.begin(), bins.end());
+	}
+	else // bvar_count > 2
+	{
+		for(int j=0; j<bvars.size()-1; j++)
+		{
+			Expr lt = mk<LT>(bvars[j], bvars[j+1]);
+			bins.push_back(lt);
+		}
+		cand = mknary<AND>(bins.begin(), bins.end());
+	}
+
+	return cand;
+  }
+
+  #define SAT true
+  #define UNSAT false
+
+  std::map<Expr, Expr> Houdini::currentCandidates;	//a map from predicates to candidates
+
+  /*
+   * Build the map from predicates to candidates
+   */
+  void Houdini::guessCandidate(HornClauseDB &db)
+  {
+	  for(Expr rel : db.getRelations())
+	  {
+		  if(bind::isFdecl(rel))
+		  {
+			  Expr cand = relToCand(rel);
+			  currentCandidates.insert(std::pair<Expr, Expr>(rel, cand));
+		  }
+	  }
+  }
+
+  /*
+   * Main loop of Houdini algorithm
+   */
+  void Houdini::runHoudini(HornClauseDB &db, HornifyModule &hm)
+  {
+	  guessCandidate(db);
+
+	  HornClauseDB::RuleVector workList;
+	  workList.insert(workList.end(), db.getRules().begin(), db.getRules().end());
+
+	  while(!workList.empty())
+	  {
+		  HornRule r = workList.front();
+		  errs() << "RULE HEAD: " << *(r.head()) << "\n";
+		  errs() << "RULE BODY: " << *(r.body()) << "\n";
+		  while (validateRule(r, db, hm) == SAT)
+		  {
+			  workList = addUsedRulesBackToWorkList(db, workList, r.head());
```

Comment:
Yes, but you are updating it inside the function already.

On Tue, Jun 28, 2016 at 15:11 Chenguang-Zhu <notifications@github.com>
wrote:

> In lib/seahorn/Houdini.cc
> <https://github.com/seahorn/seahorn/pull/33#discussion_r68822771>:
>
> > +   */
> > +  void Houdini::runHoudini(HornClauseDB &db, HornifyModule &hm)
> > +  {
> > +	  guessCandidate(db);
> > +
> > +	  HornClauseDB::RuleVector workList;
> > +	  workList.insert(workList.end(), db.getRules().begin(), db.getRules().end());
> > +
> > +	  while(!workList.empty())
> > +	  {
> > +		  HornRule r = workList.front();
> > +		  errs() << "RULE HEAD: " << *(r.head()) << "\n";
> > +		  errs() << "RULE BODY: " << *(r.body()) << "\n";
> > +		  while (validateRule(r, db, hm) == SAT)
> > +		  {
> > +			  workList = addUsedRulesBackToWorkList(db, workList, r.head());
>
> I'm a little confused about this comment, shouldn't we update the workList
> itself?
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/seahorn/seahorn/pull/33/files/cce7a302670f7c9e1a2207c15b3c5d1ac93ec881#r68822771>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AFzN5_M52WJl1g8L3XmKPdwNMGukfYXWks5qQXHbgaJpZM4I7V-j>
> .
>